### PR TITLE
plan: rewrite "hashCode()" for "requiredProp"

### LINKED ISSUE #4439: `requiredProp.getHashKey()` takes a lot of time
This function firstly makes a slice of `Datum`, then calls function `codec.EncodeValue()`.
There is only 4 value types in the `Datum` slice: `bool, string, int, float64`, maybe we can write a specific hash function for struct `requiredProp`

### PR DESCRIPTION
**Pull request by zz-jason**
fix #4439 

```go
package plan

import (
        "testing"

        "github.com/pingcap/tidb/expression"
)

func createProp() *requiredProp {
        col1 := &expression.Column{
                FromID:   1,
                Position: 12,
        }
        col2 := &expression.Column{
                FromID:   11,
                Position: 2,
        }

        return &requiredProp{
                desc:        true,
                taskTp:      12,
                expectedCnt: 1.998,
                cols:        []*expression.Column{col1, col2},
        }
}

func BenchmarkHashCodeNew(b *testing.B) {
        p := createProp()

        b.ResetTimer()
        for i := 0; i < b.N; i++ {
                p.hashcode = nil
                _ = p.hashCode()
        }
}

func BenchmarkHashCodeOrigin(b *testing.B) {
        p := createProp()

        b.ResetTimer()
        for i := 0; i < b.N; i++ {
                _, _ = p.getHashKey()
        }
}
```

```sh
$go test -bench=. -run=BenchmarkHashCode
BenchmarkHashCodeNew-8          20000000                67.9 ns/op
BenchmarkHashCodeOrigin-8        2000000               675 ns/op
PASS
ok      github.com/pingcap/tidb/plan    3.503s
```

### CODE REVIEW

**Comment by coocood** [comment]
LGTM

File: plan/plan.go [hunk 1]
```diff
@@ -296,12 +305,9 @@ func (bp *basePhysicalPlan) ExplainInfo() string {
 	return ""
 }
 
-func (p *baseLogicalPlan) getTask(prop *requiredProp) (task, error) {
-	key, err := prop.getHashKey()
```
Comment (XuHuaiyu) [inline plan/plan.go]: getHashKey can be removed?

Comment (zz-jason) [reply plan/plan.go]: already removed.

**Pull request merged**